### ISSUE TITLE
'receive' matcher doesn't support custom failure messages

### ISSUE THREAD

**Issue by rlue** [rspec/rspec-mocks#1250]
### Subject of the issue

`rspec-expectations` supports the [passing of customized failure messages to the `#to` method](https://relishapp.com/rspec/rspec-expectations/docs/customized-message):

```ruby
expect(array).to be_empty, "expected empty array, got #{array.inspect}"
```

The documentation states that

> This works for any matcher _other than the operator matchers._

This does not work for the `receive` matcher, however, and there's no mention of it in the `rspec-mocks` documentation.

### Your environment

* Ruby version: 2.4.4
* rspec-mocks version: 3.8.0

### Steps to reproduce

```ruby
RSpec.describe RSpec do
  # rspec outputs this custom failure message
  it 'supports custom failure messages for base matchers' do
    expect(1).to eq(2), 'something went wrong!'
  end

  # rspec does NOT output this custom failure message
  it 'supports custom failure messages for rspec-mocks matchers' do
    expect(RSpec).to receive(:describe), 'something went wrong!'
  end
end
```

### Expected behavior

I expect the second example to output the custom message, `"something went wrong!"`

### Actual behavior

The default failure message is output instead:

```
(RSpec).describe(*(any args))
    expected: 1 time with any arguments
    received: 0 times with any arguments
```

**Comment by benoittgt** [comment]
I can have a look at this one.

**Comment by JonRowe** [comment]
@benoittgt It will be to do with how it is passed around, this is not a technically supported feature of these matchers (unlike expectations) but I feel like this should work for the "principle of least surprise"

**Comment by benoittgt** [comment]
> but I feel like this should work for the "principle of least surprise"
Yes I have same feeling

**Comment by benoittgt** [comment]
I have started working on this one. I think I understand a little bit where is the issue. `receive` properly get the custom message but it seems it is not able to handle it the way we display custom failure message.

```ruby
 matcher.matches?(actual, &block) || ExpectationHelper.handle_failure(matcher, message, :failure_message)
```
([source](https://github.com/rspec/rspec-expectations/blob/861355ec3b0654ed12147190a5d54fba2553bcef/lib/rspec/expectations/handler.rb#L49-L50))

In case of `receive` matcher it act differently on `matcher.matches?(actual, &block)`. `have_received` doesn't have this issue.

`matcher.matches?(actual, &block)` can't properly answer in `handle_matcher` because `receive` is not fully *executed*.

Example:
```ruby
  it 'display custom mock message' do
    expect(1).to eq(3), 'custom eq message'
  end

  it do
    expect(RSpec).to have_receive(:describe), '1 have_receive message'
  end

  it do
    expect(RSpec).to receive(:describe), '2 have_receive message'
  end
```

With
```diff
       def self.handle_matcher(actual, initial_matcher, message=nil, &block)
         ExpectationHelper.with_matcher(self, initial_matcher, message) do |matcher|
           return ::RSpec::Matchers::BuiltIn::PositiveOperatorMatcher.new(actual) unless initial_matcher
+          puts "_" * 60
+          puts "lib/rspec/expectations/handler.rb:53"
+          p message
+          puts "matcher.matches?(actual, &block) ="
+          p matcher.matches?(actual, &block)
+          puts "-" * 60
```

You get

```
____________________________________________________________
lib/rspec/expectations/handler.rb:53
"custom eq message"
matcher.matches?(actual, &block) =
false
------------------------------------------------------------
F____________________________________________________________
lib/rspec/expectations/handler.rb:53
"1 have_receive message"
matcher.matches?(actual, &block) =
false
------------------------------------------------------------
F____________________________________________________________
lib/rspec/expectations/handler.rb:53
"2 have_receive message"
matcher.matches?(actual, &block) =
#<RSpec::Mocks::MessageExpectation RSpec.describe(any arguments)>
------------------------------------------------------------
F
```

I will keep trying to find a solution. But if you have an idea @JonRowe?

**Comment by JonRowe** [comment]
So it needs to be bundled up and deferred to be used later?

**Comment by benoittgt** [comment]
Yes @JonRowe. I try to look at it a little bit more and find that you can set a custom message using this `opts[:message]` in https://github.com/rspec/rspec-mocks/blob/7dc2a431101172d31b780458eb168e05d20fb3aa/lib/rspec/mocks/error_generator.rb#L308. I need to understand how I could use it I will try this week but I am so busy. 😟

**Comment by ronilbhatia** [comment]
Any update on this issue? I am currently facing the same problem, and unable to figure out how to set a custom failure message for `receive`. Would there be a temporary workaround such as monkeypatching `RSpec::Mocks::Matchers::Receive` that could provide a solution in the interim?

**Comment by rlue** [comment]
@benoittgt, thanks so much for your insight! It really saved me a lot of time in wrapping my head around what's happening here.

### Problem summary

1. Custom error messages are [given to `ExpectationTarget#to`](https://github.com/rspec/rspec-expectations/blob/861355ec3b0654ed12147190a5d54fba2553bcef/lib/rspec/expectations/expectation_target.rb#L63-L66), as in:

   ```ruby
   expect(this).to(eq(that), 'or else print this message')
   ```

2. `#to` passes the message to `PositiveExpectationHandler.handle_matcher`, which [asks the matcher if it succeeded and, if not, passes the message yet again to `ExpectationHelper.handle_failure`](https://github.com/rspec/rspec-expectations/blob/861355ec3b0654ed12147190a5d54fba2553bcef/lib/rspec/expectations/handler.rb#L49-L50).

3. Step 2 fails for the `receive` matcher, because `receive` can't tell whether it's succeeded or not until after the rest of the example has run. So instead, it bypasses this check and provides its own failure-handling logic. 

   Trying to get a custom failure message into `receive`'s _ad hoc_ failure-handling logic is tricky, because the message doesn't live in the matcher object itself—it lives in the `#to` method, which has long since concluded by the time `receive` needs it.

### Proposed solutions

Please correct me if I sound naive—you guys definitely have way more experience than I do—but it seems to me that the way that the `receive` matcher is written makes it hard to solve this problem in an open–closed fashion.

1. @JonRowe & co., what would you think of an alternative matcher to `receive`, where the test code is wrapped in a block?

   ```ruby
   expect { foo }.to invoke(obj, :method)
   ```

   Would this even solve the problem? If you guys are open to this, I'm more than willing to work on a PR.

2. Alternately, what if we:

   1. added a `:failure_message` attribute to all matchers,
   2. rewrote `#to` to set it when a custom failure message is given, and
   3. rewrote `.handle_failure` to use it instead of taking the message as an argument? 

   Then, the `receive` matcher would store its own failure message in an instance variable, so it could reference it when needed.

**Comment by benoittgt** [comment]
Hello

I did some pairing this morning with @nicoolas25 on this issue. We wrote draft patchs:
- https://github.com/rspec/rspec-expectations/pull/1156
- https://github.com/rspec/rspec-mocks/pull/1312

With:
```ruby
require 'spec_helper'

RSpec.describe "custom message" do
  it 'display custom mock message' do
    expect(1).to eq(3), 'CUSTOM eq message'
  end

  it do
    expect(RSpec).to receive(:describe), 'CUSTOM receive message'
  end
end
```


It produces:

```
$ rspec  
FF

Failures:

  1) custom message display custom mock message
     Failure/Error: expect(1).to eq(3), 'CUSTOM eq message'
       CUSTOM eq message
     # ./spec/repro_spec.rb:5:in `block (2 levels) in <top (required)>'

  2) custom message is expected to receive describe(*(any args)) 1 time
     Failure/Error: expect(RSpec).to receive(:describe), 'CUSTOM receive message'
       CUSTOM receive message
     # ./spec/repro_spec.rb:9:in `block (2 levels) in <top (required)>'

Finished in 0.02302 seconds (files took 0.1084 seconds to load)
2 examples, 2 failures

Failed examples:

rspec ./spec/repro_spec.rb:4 # custom message display custom mock message
rspec ./spec/repro_spec.rb:8 # custom message is expected to receive describe(*(any args)) 1 time
```

----

Is something like that you were thinking @JonRowe in https://github.com/rspec/rspec-mocks/issues/1250#issuecomment-453199176?

We try to do minimal changes but the result in https://github.com/rspec/rspec-expectations/pull/1156 looks maybe a little bit "hacky".

Thanks in advance for the review. I will add tests when the idea will be validated.

**Comment by JonRowe** [comment]
Something like this looks good, I've only quickly scanned it this morning and I suspect I'd have some points for a proper review but the approach doesn't look wrong?

**Comment by benoittgt** [comment]
Great! I will add specs and will turn draft into ready-to-review PR. :)

**Comment by benoittgt** [comment]
PR are ok. Should we add doc? What about in https://relishapp.com/rspec/rspec-mocks/v/3-9/docs/basics/expecting-messages. I am not sure. It seems a little bit out of context.

**Comment by JonRowe** [comment]
Yes I would add docs somewhere, an additional scenario mentioning you can customise the failure message...

**Comment by benoittgt** [comment]
A demo can be seen here https://relishapp.com/rspec-staging/rspec-mocks/docs/basics/customized-message

Commit: https://github.com/rspec/rspec-mocks/pull/1312/commits/a7a2923e3c8671a6f1ee0ec1de79fb3eb19ba899

**Comment by pirj** [comment]
Maybe move it as additional scenarios to https://relishapp.com/rspec-staging/rspec-mocks/docs/basics/expecting-messages instead?

**Comment by benoittgt** [comment]
Good idea @pirj

**Comment by benoittgt** [comment]
Here it is https://relishapp.com/rspec-staging/rspec-mocks/docs/basics/expecting-messages

**Comment by JonRowe** [comment]
Much better in the second link

**Comment by dmlary** [comment]
Any chance this could be added to a point release for the gem?  I'm working with some tests that have illegible arguments; including the custom message would make them so much easier to debug.

**Comment by JonRowe** [comment]
@dmlary I'm afraid not, as per our semantic versioning policy this is an enhancement and will only be released in a minor release, as this a complicated matter for us due to having several gems to coordinate minor releases of, we tend to only release them when we have sufficient things to release, however as it's been a few months, I'll try to get one done sooner rather than later, in the mean time you can use the github repositories directly.

**Comment by dmlary** [comment]
Thank you! I completely forgot I could just point the Gemfile at the github repo.
